plugins/catalog-backend/src/service/createRouter.ts

```diff
@@ -135,24 +138,94 @@ export async function createRouter(
+
+        if (pagination) {
+          const { entities, pageInfo } = await entitiesCatalog.entities({
+            filter,
+            fields,
+            order,
+            pagination,
+            credentials,
+          });
+
+          // Add a Link header to the next page
+          if (pageInfo.hasNextPage) {
+            const url = new URL(`http://ignored${req.url}`);
+            url.searchParams.delete('offset');
+            url.searchParams.set('after', pageInfo.endCursor);
+            res.setHeader('link', `<${url.pathname}${url.search}>; rel="next"`);
+          }
 
-        // Add a Link header to the next page
-        if (pageInfo.hasNextPage) {
-          const url = new URL(`http://ignored${req.url}`);
-          url.searchParams.delete('offset');
-          url.searchParams.set('after', pageInfo.endCursor);
-          res.setHeader('link', `<${url.pathname}${url.search}>; rel="next"`);
+          res.json(entities);
+          return;
         }
 
-        // TODO(freben): encode the pageInfo in the response
-        res.json(entities);
+        // For other read-the-entire-world cases, use queryEntities and stream
+        // out results.
+
+        // The write lock is used for back pressure, preventing slow readers
+        // from forcing our read loop to pile up response data in userspace
+        // buffers faster than the kernel buffer is emptied.
+        // https://nodejs.org/api/http.html#http_response_write_chunk_encoding_callback
+        let writeLock: DeferredPromise | undefined;
+        const controller = new AbortController();
+        const signal = controller.signal;
+        req.on('end', () => {
+          controller.abort(new Error('Client closed connection'));
+          writeLock?.resolve();
+          writeLock = undefined;
+        });
+        res.on('drain', () => {
+          writeLock?.resolve();
+          writeLock = undefined;
+        });
+
+        const responseStream = createEntityArrayJsonStream(res);
+        const limit = 10000;
+        let cursor: Cursor | undefined;
+
+        try {
+          do {
+            const result = await entitiesCatalog.queryEntities(
+              !cursor
+                ? { credentials, fields, limit, filter, orderFields: order }
+                : { credentials, fields, limit, cursor },
+            );
+
+            if (result.items.length) {
+              if (writeLock) {
+                await writeLock;
+              }
+
+              signal.throwIfAborted();
+
+              expandLegacyCompoundRelationRefsInResponse(result.items);
+              if (!responseStream.send(result.items)) {
+                // The kernel buffer is full. Create the lock but do not await it
+                // yet - we can better spend our time going to the next round of
+                // the loop and read from the database while we wait for it to
+                // drain.
+                writeLock = createDeferred();
```

Comment:
I wonder if doing `res.once('drain', () => { });` and somehow awaiting for the drain event could simplify things 🤔

Reply:
Updated!